core/types/block.go

```diff
@@ -603,6 +654,115 @@ func (h *Header) EmptyReceipts() bool {
+		return totalS
+	case common.REGION_CTX:
+		totalS := new(big.Int).Add(h.ParentEntropy(common.PRIME_CTX), h.ParentDeltaS(common.REGION_CTX))
+		return totalS
+	case common.ZONE_CTX:
+		totalS := new(big.Int).Add(h.ParentEntropy(common.PRIME_CTX), h.ParentDeltaS(common.REGION_CTX))
+		totalS.Add(totalS, h.ParentDeltaS(common.ZONE_CTX))
+		return totalS
+	}
+	return big.NewInt(0)
+}
+
+func (h *Header) CalcDeltaS() *big.Int {
+	order := h.CalcOrder()
+	intrinsicS := h.CalcIntrinsicS()
+	switch order {
+	case common.PRIME_CTX:
+		return big.NewInt(0)
+	case common.REGION_CTX:
+		totalDeltaS := new(big.Int).Add(h.ParentDeltaS(common.REGION_CTX), h.ParentDeltaS(common.ZONE_CTX))
+		totalDeltaS = new(big.Int).Add(totalDeltaS, intrinsicS)
+		return totalDeltaS
+	case common.ZONE_CTX:
+		totalDeltaS := new(big.Int).Add(h.ParentDeltaS(common.ZONE_CTX), intrinsicS)
+		return totalDeltaS
+	}
+	return big.NewInt(0)
+}
+
+func (h *Header) CalcOrder() int {
+	intrinsicS := h.CalcIntrinsicS()
+
+	// This is the updated the threshold calculation based on the zone difficulty threshold
+	zoneThresholdS := h.CalcIntrinsicS(common.BytesToHash(new(big.Int).Div(big2e256, h.Difficulty()).Bytes()))
+	timeFactorHierarchyDepthMultiple := new(big.Int).Mul(params.TimeFactor, big.NewInt(common.HierarchyDepth))
+
+	// Prime case
+	primeEntropyThreshold := new(big.Int).Mul(timeFactorHierarchyDepthMultiple, timeFactorHierarchyDepthMultiple)
+	primeEntropyThreshold = new(big.Int).Mul(primeEntropyThreshold, zoneThresholdS)
+	primeBlockThreshold := new(big.Int).Quo(primeEntropyThreshold, big.NewInt(10))
+	primeEntropyThreshold = new(big.Int).Sub(primeEntropyThreshold, primeBlockThreshold)
+
+	primeBlockEntropyThresholdAdder, _ := mathutil.BinaryLog(primeBlockThreshold, 8)
+	primeBlockEntropyThreshold := new(big.Int).Add(zoneThresholdS, big.NewInt(int64(primeBlockEntropyThresholdAdder)))
+
+	totalDeltaS := new(big.Int).Add(h.ParentDeltaS(common.REGION_CTX), h.ParentDeltaS(common.ZONE_CTX))
+	totalDeltaS.Add(totalDeltaS, intrinsicS)
+	if intrinsicS.Cmp(primeBlockEntropyThreshold) > 0 && totalDeltaS.Cmp(primeEntropyThreshold) > 0 {
+		return common.PRIME_CTX
+	}
+
+	// Region case
+	regionEntropyThreshold := new(big.Int).Mul(timeFactorHierarchyDepthMultiple, zoneThresholdS)
+	regionBlockThreshold := new(big.Int).Quo(regionEntropyThreshold, big.NewInt(10))
+	regionEntropyThreshold = new(big.Int).Sub(regionEntropyThreshold, regionBlockThreshold)
+
+	regionBlockEntropyThresholdAdder, _ := mathutil.BinaryLog(regionBlockThreshold, 8)
+	regionBlockEntropyThreshold := new(big.Int).Add(zoneThresholdS, big.NewInt(int64(regionBlockEntropyThresholdAdder)))
+
+	totalDeltaS = new(big.Int).Add(h.ParentDeltaS(common.ZONE_CTX), intrinsicS)
+	if intrinsicS.Cmp(regionBlockEntropyThreshold) > 0 && totalDeltaS.Cmp(regionEntropyThreshold) > 0 {
+		return common.REGION_CTX
+	}
+
+	// Zone case
+	if intrinsicS.Cmp(zoneThresholdS) > 0 {
+		return common.ZONE_CTX
+	}
+	return -1
```

Comment:
Being changed to return error instead of -1 as sentinel

Reply:
adding error and catching it